nemo/utils/callbacks/preemption.py

```diff
@@ -91,10 +91,8 @@ def on_train_batch_end(self, trainer, pl_module, outputs, batch, batch_idx: int)
             # a regular local variable
             interrupted = self.interrupted
             if interrupted:
-                logging.info("Received SIGTERM, saving checkpoint and exiting")
-                monitor_candidates = self.checkpoint_callback._monitor_candidates(trainer)
-                self.checkpoint_callback._save_last_checkpoint(trainer, monitor_candidates)
-                sys.exit(0)
+                logging.info("Preemption detected, signaling trainer to stop")
+                trainer.should_stop = True
```

Comment:
@jbieniusiewi how is the last checkpoint saved in this case ?

Reply:
In my experiments, when should_stop=True, it consistently: runs the evaluation loop and saves the checkpoint before terminating.  I'm not sure if there are some PTL flags that control this behaviour.

@cuichenx I've noticed that you've added a new preemption callback that stops the training with `trainer.should_stop=True`. Could you please share your experience with this? Specifically, does the checkpoint get saved when `trainer.should_stop` is set to True?

Reply:
I'm closing this PR, due to the possible problems with the last checkpoint saving.

Reply:
Hi @jbieniusiewi I don't think I added it.. which one are you referring to?

Reply:
Hi @cuichenx 

By the "new preemption callback" I meant: `nemo/lightning/pytorch/callbacks/preemption.py`

[Which uses "trainer.should_stop=True" to stop the training.](https://github.com/NVIDIA/NeMo/blob/68d4dba331fdafbb33a640f4e3ed20fc660864a7/nemo/lightning/pytorch/callbacks/preemption.py#L65)

But it seems it has been recently fixed [on "r2.0.1" branch](https://github.com/NVIDIA/NeMo/blob/da0b0b7cae62358f2023851fd04ac90d123bc159/nemo/lightning/pytorch/callbacks/preemption.py#L72)

And the updated version uses this construct to save the checkpoint and stop the training:

```
  monitor_candidates = trainer.checkpoint_callback._monitor_candidates(trainer)
  trainer.checkpoint_callback._save_last_checkpoint(trainer, monitor_candidates)
  sys.exit(0)
```

So I infer that using "trainer.should_stop=True" is not correct.

Reply:
@jbieniusiewi , @hemildesai added it in `nemo/lightning/pytorch/callbacks/preemption.py`. He should be able to provide comments for that.